datumaro/cli/util/project.py

```diff
@@ -22,17 +28,104 @@ def generate_next_file_name(basename, basedir='.', sep='.', ext=''):
 
     return generate_next_name(os.listdir(basedir), basename, sep, ext)
 
-def generate_next_name(names, basename, sep='.', suffix='', default=None):
-    pattern = re.compile(r'%s(?:%s(\d+))?%s' % \
-        tuple(map(re.escape, [basename, sep, suffix])))
-    matches = [match for match in (pattern.match(n) for n in names) if match]
+_dataset_revpath_regex = None
+_full_revpath_regex = None
+def parse_full_revpath(s: str, ctx_project: Optional[Project]):
+    if ctx_project:
+        env = ctx_project.env
+    else:
+        env = Environment()
+
+    def parse_dataset_pathspec(s: str):
+        global _dataset_revpath_regex
+        if not _dataset_revpath_regex:
```

Comment:
Are you sure it's worthwhile to implement this lazy initialization scheme? I would've just moved the initialization to the global level.

Reply:
Actually, it looks like Python caches regexes itself. Anyway, caching or precompiling of a regex that is called one or few times gives no performance gain.

Reply:
> Actually, it looks like Python caches regexes itself.

That's true, although the cache has a limited size.

> Anyway, caching or precompiling of a regex that is called one or few times gives no performance gain.

That's also true, but doing lazy initialization gives a readability loss. Hence, you should either put the regex in a global variable and compile it immediately, or put it in a local variable.